design/005-multi-container-sessions/005-multi-container-sessions.md

```diff
@@ -0,0 +1,163 @@
+- Status: Proposed
+
+# Native Rstudio Sessions
+
+## Summary
+
+Enable Amalthea to support defining more than one container that is exposed and accessible not
+just the regular /lab endpoint. In the case of Rstudio currently we support both /lab and /studio
+but the proxying for this is done at the Jupyterlab level. The idea here is that such proxying
+should be done by Amalthea.
+
+## Motivation
+
+- Users can run additional programs like dashboards beside their regular session
+- We can run alternative UIs like vscode or rstudio alongside jupyterlab without relying on the proxying
+from rstudio
+- We further generalize Amalthea to accept any type of service/container
+
+## Design Detail
+
+### Current manifest
+
+```yaml
+spec:
+  culling:
+    idleSecondsThreshold: 0
+    maxAgeSecondsThreshold: 0
+  auth:
+    oidc:
+      authorizedEmails:
+      - <user-email>
+      authorizedGroups: []
+      clientId: OidcClientName
+      clientSecret:
+        value: OidcSecretValue
+      enabled: true
+      issuerUrl: https://dev.renku.ch/auth/realms/Renku
+    token: ""
+  jupyterServer:
+    defaultUrl:
+    image:
+    resources:
+    rootDir: /home/jovyanl/work/<project_name>
+  routing:
+    host: dev.renku.ch
+    ingressAnnotations:
+      kubernetes.io/ingress.class: nginx
+      nginx.ingress.kubernetes.io/proxy-body-size: "0"
+      nginx.ingress.kubernetes.io/proxy-buffer-size: 8k
+      nginx.ingress.kubernetes.io/proxy-request-buffering: "off"
+    path: /sessions/<session_name>
+    tls:
+      enabled: true
+      secretName: dev-renku-ch-tls
+  storage:
+    pvc:
+      enabled:
+      mountPath: /home/jovyanl/work
+      storageClassName:
+    size:
+```
+
+### Proposed revised manifest
+
+```yaml
+spec:
+  template:
+    configMap:
+      name: jupyter-configmap
+  culling:
+    idleSecondsThreshold: 0
+    maxAgeSecondsThreshold: 0
+    idleProbe: # NEW similar syntax expected to http get k8s probe
+  auth:
+    oidc:
+    token:
+  servers: # NEW
+    - containerName: jupyter-server
+      image:
+      resources:
+        requests:
+          memory: 1G
+          cpu: 1
+      ingress:
+        - port: 8889
+          url: /lab
+      livenessProbe: {}
+      readinessProbe: {}
+      startupProbe: {}
+    - containerName: dashboard
+      image:
+      resources:
+        requests:
+          memory: 1G
+          cpu: 1
+      ingress:
+        - port: 8890
+          url: /dashboard
+      livenessProbe: {}
+      readinessProbe: {}
+      startupProbe: {}
+  routing:
+    host: dev.renku.ch
+    ingressAnnotations:
+      kubernetes.io/ingress.class: nginx
+      nginx.ingress.kubernetes.io/proxy-body-size: "0"
+      nginx.ingress.kubernetes.io/proxy-buffer-size: 8k
+      nginx.ingress.kubernetes.io/proxy-request-buffering: "off"
+    path: /sessions/<session_name>
+    tls:
+      enabled: true
+      secretName: dev-renku-ch-tls
+  storage:
+    pvc:
+      enabled:
+      mountPath: /home/jovyanl/work
```

Comment:
```suggestion
      mountPath: /home/jovyan/work
```

the `mountPath` might need to be set per container?

Reply:
I was thinking that we just mount the pvc in the same place on all containers. This is the spec for the jupyterserver so it does not have to be a valid pod spec.

I just thought it was easier to enforce this rather than to have freedom to mount things in different locations.

Reply:
But it is definitely doable to have more freeedom and mount thing differently across different "apps"/containers.

Reply:
The problem is that this assumes a jovyan user exists - which potentially complicates the requirements of the image. But maybe it's a non-issue? not sure.

Reply:
Ok yeah I will add this to the questions that should be resolved then.

It will be nice to not have to worry about this.

Reply:
still, "jovyan" is misspelled ;)